### PR TITLE
Draft of SDLC

### PR DESCRIPTION
**Pull request by mbryzek**

### CODE REVIEW

File: Sdlc.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
+# Software Development Lifecycle
+
+
+Flow's Software Development Lifecycle (SDLC) is focused on creating an amazing environment for individuals and teams to contribute safely to the growth of Flow. One of the pillars of our culture is trust and empowerment - and our SDLC is informed by and designed to demonstate and further this value.
```
Comment (mikeclimbrock) [inline Sdlc.md]: "the growth of Flow" => "Flow's growth"

Comment (mikeclimbrock) [inline Sdlc.md]: "One of the pillars" => "One pillar"

File: Sdlc.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+## Philosophy
+
+We pursue several practices to ensure our SDLC achieves its goals of providing very high productivity for each of us while continually reducing risk.
```
Comment (mikeclimbrock) [inline Sdlc.md]: remove "for each of us"

File: Sdlc.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+  - We remind ourselves constantly that **failure is a mandatory component of success**. As people, learning and failure simply go hand in hand.
```
Comment (mikeclimbrock) [inline Sdlc.md]: remove "simply"

File: Sdlc.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  
+  - Not all failures are created equal. We use the word **Intentionality** to describe an intentional change to a system. In particular, failures that result from intentional changes are the changes we expect to learn from; by contrast, a lack of intentionality leads to negligence or carelessness which erodes trust and therefore is contrary to our culture.
+  
+  - We **learn and adapt from prior learnings**; In particular, from Kaizen we adopt the learnings of continuous improvement. When failure occurs, we immediately determine what needs to be done to prevent recurrence and focus on those tasks. We learn about antifragile - the notion that following each learning, we emerge stronger as an organization for the future. We learn from Scrum and Cambam ways to organize our work; but we also learn their shortcoming and adjust appropriately for our environment.
```
Comment (mikeclimbrock) [inline Sdlc.md]: add a "," after Kaizen

Comment (mikeclimbrock) [inline Sdlc.md]: "Cambam" => "Kanban"

File: Sdlc.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+  
+  - We focus on software and architecture that continually reduces the risk of changes. These practices focus on:
+  
+    - isolation - truly ensuring that one change cannot impact another. For example, our focus on API design and micro service architecture on the backend is one technique we use to create isolation
```
Comment (mikeclimbrock) [inline Sdlc.md]: some of these bullets have periods, and others do not.  make consistent

File: Sdlc.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    
+    - continuous delivery - by releasing small changes continuously to our production environment, we can minimze the impact of any one change and significantly increase time to resolution.
```
Comment (mikeclimbrock) [inline Sdlc.md]: remove "can" => "we minimize"

File: Sdlc.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    
+    - automation - by writing software to manage and implement our processes - including testing, patching and deployment - we focus on capturing our assumptions during the development process and building leverage to ensure our assumptions remain valid over time
+    
+    - rollback - by ensuring that every application change can be rolled back, we enable faster resolution when problems arise. For example, our practice of managing database schema changes independently of application changes provides the means to ensure every application change can itself be rolled back.
+    
+    - incremental rollout - by slowly deploying application changes to production, and verifying that they work, we can minimize the impact of a problem before it affects all of our users
```
Comment (mikeclimbrock) [inline Sdlc.md]: now sure "slowly" is correct word here.

File: Sdlc.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    
+    - monitoring and alerting - by instrumenting our applications - via logging and external monitoring services - we can capture our assumptions and receive notification on variance to minimize customer impact.
+   
+## Application Development at Flow
+
+### Design
+
+Design of the system starts as soon as we begin discussing use cases and requirements. While the design stage starts early in our process, it is not concluded until the application is successfully completed - simple, stable, tested, in production, and meeting the needs of users.
+
+APIs designed for our customers further must meet the requirements of the 'api' repository - https://github.com/flowcommerce/api - and designed using [apidoc](apidoc.me/flow)
```
Comment (mikeclimbrock) [inline Sdlc.md]: "further must meet" => "must additionally meet"

File: Sdlc.md [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+### Environments
+
+Flow provides three environments:
+
+1. Local development environment
+2. Local integration environment, supported by [workstation](https://github.com/flowcommerce/workstation)
+3. Production, supported by
```
Comment (mikeclimbrock) [inline Sdlc.md]: missing link to delta? [delta](https://github.com/flowcommerce/delta)

File: Sdlc.md [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+### Environments
+
+Flow provides three environments:
+
+1. Local development environment
+2. Local integration environment, supported by [workstation](https://github.com/flowcommerce/workstation)
+3. Production, supported by 
+
+In all cases, docker artifacts must be produced as these are the artifacts used in both the integration and production environments.
```
Comment (mikeclimbrock) [inline Sdlc.md]: "are the artifacts" "artifacts are"

File: Sdlc.md [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+  - [Source code management](https://github.com/flowcommerce)
+  - [Contributions](https://github.com/flowcommerce/standards/Contributions.md)
+  
+For top secret applications:
+
+  - [Source code management](https://github.com/flowvault)
+  - [Contributions](https://github.com/flowvault/standards/Contributions.md)
+  
+We also have decided as a group that we prefer our applications to be developed with similar style - thus making it easy for a developer to move from one project to another. Thus you will see very similar style of development across all Play/Scala applications at Flow as you will see across all NodeJs applications.
```
Comment (mikeclimbrock) [inline Sdlc.md]: "We also have decided as a group" => "As a group, we decided"

Comment (mikeclimbrock) [inline Sdlc.md]: "similar style" => "similar styles"

Comment (mikeclimbrock) [inline Sdlc.md]: "as you will see across" => "and across"

File: Sdlc.md [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+### Dependendency Management
+
+We strive to minimize the number of external dependencies. When we do bring in dependencies, we are resposible for reviewing the code as if it is our own - and we remain accountable for its performance and security at Flow.
```
Comment (mikeclimbrock) [inline Sdlc.md]: "When we do bring in dependencies" => "When new dependencies are introduced"

File: Sdlc.md [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+
+To consume third party APIs - our preference, where possible, is to first document the API using [apidoc](apidoc.me/flow) which then ensures we can consume the third party API without introducing new dependencies.
+
+To track third party dependencies, for Scala based applications, we use [dependency](dependency.flow.io) and for Node applications we use grasshopper. We commit to "OCD Fridays" where we spend the morning every week upgrading to the latest versions of all libraries. This practice of continuous upgrade aligns with our philosophy of continuous delivery.
```
Comment (mikeclimbrock) [inline Sdlc.md]: add "," after "Node applications,"

File: Sdlc.md [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+
+### Deployment
+
+In non Top Secret environments, deployment is via [delta](https://github.com/flowcommerce/delta) - available at [https://delta.flow.io](https://delta.flow.io)
+
+In Top Secret environmnents, refer to the deployment instructions for each application for details on secure deployment.
+
+All infrastructure runs on AWS, preferring the use of their shared services for storage (e.g. RDS).
+
+### Erors
```
Comment (mikeclimbrock) [inline Sdlc.md]: "Errors"

File: Sdlc.md [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+
+When errors occur, teams should immediately respond:
+
+  - We always move forward; if necessary rollback by creating a new git tag
```
Comment (mikeclimbrock) [inline Sdlc.md]: add "," after "if necessary,"

File: Sdlc.md [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+  - For any non trivial error, announce immediately in slack channel and make sure the people that can help are helping (e.g. interrupt any discussion or meeting, call them, etc)
```
Comment (mikeclimbrock) [inline Sdlc.md]: "slack" => "Slack"

**Comment by mikeclimbrock** [comment]
- should there be any mention of adding an application to Registry?
- should there be any mention to configure integration builds in TravisCI (.com vs. .org)
- anything on publishing new artifacts to Artifactory?
- mention weekly standup?

File: Sdlc.md (hunk 5 above)
Comment (chrhicks) [inline Sdlc.md]: add periods.. you're actually writing full sentences.

File: Sdlc.md (hunk 7 above)
Comment (chrhicks) [inline Sdlc.md]: incrementally ?

File: Sdlc.md [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+  - [Source code management](https://github.com/flowvault)
+  - [Contributions](https://github.com/flowvault/standards/Contributions.md)
+  
+We also have decided as a group that we prefer our applications to be developed with similar style - thus making it easy for a developer to move from one project to another. Thus you will see very similar style of development across all Play/Scala applications at Flow as you will see across all NodeJs applications.
+
+### Dependendency Management
+
+We strive to minimize the number of external dependencies. When we do bring in dependencies, we are resposible for reviewing the code as if it is our own - and we remain accountable for its performance and security at Flow.
+
+To consume third party APIs - our preference, where possible, is to first document the API using [apidoc](apidoc.me/flow) which then ensures we can consume the third party API without introducing new dependencies.
```
Comment (chrhicks) [inline Sdlc.md]: remove `is to first document` -> `is to document`

File: Sdlc.md (hunk 13 above)
Comment (chrhicks) [inline Sdlc.md]: grasshopper - we use `npm` , specifically `npm outdated` now

File: Sdlc.md (hunk 1 above)
Comment (mbryzek) [inline Sdlc.md]: added

File: Sdlc.md (hunk 2 above)
Comment (mbryzek) [inline Sdlc.md]: done

File: Sdlc.md (hunk 3 above)
Comment (mbryzek) [inline Sdlc.md]: done

File: Sdlc.md (hunk 4 above)
Comment (mbryzek) [inline Sdlc.md]: done

File: Sdlc.md (hunk 5 above)
Comment (mbryzek) [inline Sdlc.md]: done - all have periods

File: Sdlc.md (hunk 6 above)
Comment (mbryzek) [inline Sdlc.md]: done

File: Sdlc.md (hunk 7 above)
Comment (mbryzek) [inline Sdlc.md]: changed to by rolling out application changes to production incrementally

File: Sdlc.md (hunk 8 above)
Comment (mbryzek) [inline Sdlc.md]: chagned to must also meet

File: Sdlc.md (hunk 9 above)
Comment (mbryzek) [inline Sdlc.md]: done

File: Sdlc.md (hunk 10 above)
Comment (mbryzek) [inline Sdlc.md]: done

File: Sdlc.md (hunk 11 above)
Comment (mbryzek) [inline Sdlc.md]: done

File: Sdlc.md (hunk 12 above)
Comment (mbryzek) [inline Sdlc.md]: changed to When we introduce dependencies, we are resposible for reviewing the code as if it is our own and we remain accountable for its performance and security at Flow

File: Sdlc.md (hunk 13 above)
Comment (mbryzek) [inline Sdlc.md]: done

File: Sdlc.md (hunk 14 above)
Comment (mbryzek) [inline Sdlc.md]: done

File: Sdlc.md (hunk 15 above)
Comment (mbryzek) [inline Sdlc.md]: done

File: Sdlc.md (hunk 16 above)
Comment (mbryzek) [inline Sdlc.md]: done

File: Sdlc.md (hunk 17 above)
Comment (mbryzek) [inline Sdlc.md]: done

**Comment by mbryzek** [comment]
> should there be any mention of adding an application to Registry?
Added

> should there be any mention to configure integration builds in TravisCI (.com vs. .org)
Added

> anything on publishing new artifacts to Artifactory?
Added

> mention weekly standup?
Added

File: Sdlc.md [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+1. Local development environment
+2. Local integration environment, supported by [workstation](https://github.com/flowcommerce/workstation)
+3. Production, supported by [delta](https://delta.flow.io)
+
+To simplify local integration environments, and manage dependencies between applications, each application must be created in [registry](https://github.com/flowcommerce/registry). Registry provides the following key features:
+
+  - Unique port allocations, simplifying configuration for local integration environments
+  - Declared dependency which informs which other applications, including databases, need to be running and healthy for a given application to work
+
+Docker artifacts must be produced as these artifacts are used in both the integration and production environments. Docker artifacts are built of the master branch of each repository. Docker artifacts are stored and managed by [Docker Hub](http://hub.docker.com).
```
Comment (mikeclimbrock) [inline Sdlc.md]: typo: "of"=>"off"

**Comment by mikeclimbrock** [comment]
one more small comment, otherwise, LGTM!

**Comment by mbryzek** [comment]
fixed typo

**Pull request merged**